Add ShadowLauncherApps.getActivityList

utzcoz opened the pull request:
Test: `./gradlew clean assemble install compileTest`
Test: `./gradlew test`

utzcoz commented:
Hi @hoisie , could you help to review this pr? Thanks.

utzcoz commented:
@hoisie , thanks. I will do it.

hoisie commented on robolectric/src/test/java/org/robolectric/shadows/ShadowLauncherAppsTest.java:
```diff
@@ -27,6 +31,7 @@
+import org.robolectric.util.ReflectionHelpers;
 import org.robolectric.util.reflector.Accessor;
 import org.robolectric.util.reflector.ForType;
 
```
Can the `@Config(minSdk = O_MR1)` in this test file changed to `L`? You may have to adjust some individual tests to be `@Config(minSdk = O_MR1)`.

utzcoz replied:
The minimum sdk can used is N for test, otherwise setup will throw exception for non exist service.

utzcoz commented:
Looks like test setup can't find service when running ShadowLauncherApps with API L. I will check it locally.

hoisie commented:
I played around with this and this seems to make it work:

1) In setUp, use `launcherApps =
        (LauncherApps)ApplicationProvider.getApplicationContext().getSystemService(Context.LAUNCHER_APPS_SERVICE);` instead of using `LauncherApps.class`.
2) Make all existing tests `@Config(minSdk=O_MR1)`
3) use something like this to test L to M. The LauncherActivityInfo constructor seems to have changed in M:

http://androidxref.com/5.1.1_r6/xref/frameworks/base/core/java/android/content/pm/LauncherActivityInfo.java#52

```java
  @Test
  @Config(minSdk=L, maxSdk=M)
  public void testGetActivityListPreN() {
    String testPackageName = "test-package";
    UserHandle userHandle = UserHandle.CURRENT;

    assertThat(launcherApps.getActivityList(testPackageName, userHandle)).isEmpty();

    ResolveInfo info = ShadowResolveInfo.newResolveInfo(testPackageName, testPackageName, testPackageName);
    LauncherActivityInfo launcherActivityInfo =
        ReflectionHelpers.callConstructor(
            LauncherActivityInfo.class,
            from(Context.class, ApplicationProvider.getApplicationContext()),
            from(ResolveInfo.class, info),
            from(UserHandle.class, userHandle),
            from(long.class, System.currentTimeMillis())
        );
    shadowOf(launcherApps).addActivity(userHandle, launcherActivityInfo);
    assertThat(launcherApps.getActivityList(testPackageName, userHandle))
        .contains(launcherActivityInfo);
  }
```

utzcoz commented:
@hoisie thanks a lot.

hoisie commented:
Can you apply google-java-format on ShadowLauncherAppsTest.java?
```sh
curl -L -o $HOME/google-java-format.jar https://github.com/google/google-java-format/releases/download/google-java-format-1.8/google-java-format-1.8-all-deps.jar
curl -L -o $HOME/google-java-format-diff.py https://raw.githubusercontent.com/google/google-java-format/google-java-format-1.8/scripts/google-java-format-diff.py
git checkout support-shadowlauncherapps-getactivitylist 
git diff -U0 $(git merge-base HEAD origin/master) | python $HOME/google-java-format-diff.py --google-java-format-jar=$HOME/google-java-format.jar -p1 -i
git commit -a --amend
```

utzcoz commented:
@hoisie , after I run `git diff -U0 $(git merge-base HEAD origin/master) | python $HOME/google-java-format-diff.py --google-java-format-jar=$HOME/google-java-format.jar -p1 -i`, it changes many files out of this CL, and there are changed file list:

```
modified:   robolectric/src/test/java/org/robolectric/android/AndroidInterceptorsIntegrationTest.java
modified:   robolectric/src/test/java/org/robolectric/android/DeviceConfigTest.java
modified:   shadows/framework/src/main/java/org/robolectric/shadows/ShadowActivity.java
modified:   shadows/framework/src/main/java/org/robolectric/shadows/ShadowLocationManager.java
modified:   shadows/framework/src/main/java/org/robolectric/shadows/ShadowPausedAsyncTaskLoader.java
modified:   shadows/framework/src/main/java/org/robolectric/shadows/ShadowSmsManager.java
```
If need, I can use another CL to do it.

hoisie commented:
Ah, you probably needed to do a `git fetch origin master` to pull the latest changes.

utzcoz commented:
Thanks.
